src/qubx/backtester/simulator.py

```diff
@@ -0,0 +1,351 @@
+            start=start,
+            stop=end,
+            transform=self._transformer,
+        )
+
+        if self._timeframe:
+            args["timeframe"] = self._timeframe
+
+        return self._reader.read(**args)  # type: ignore
+
+
+class _SimulatedScheduler(BasicScheduler):
+    def run(self):
+        self._is_started = True
+        _has_tasks = False
+        _time = self.time_sec()
+        for k in self._crons.keys():
+            _has_tasks |= self._arm_schedule(k, _time)
+
+
+class SimulatedExchange(IBrokerServiceProvider):
+    _last_quotes: Dict[str, Optional[Quote]]
+    _scheduler: BasicScheduler
+    _current_time: dt_64
+    _hist_data_type: str
+    _loader: Dict[str, DataLoader]
+
+    def __init__(
+        self,
+        exchange_id: str,
+        trading_service: ITradingServiceProvider,
+        reader: DataReader,
+        hist_data_type: str = "ohlc",
+    ):
+        super().__init__(exchange_id, trading_service)
+        self._reader = reader
+        self._hist_data_type = hist_data_type
+        exchange_id = exchange_id.lower()
+
+        # - create exchange's instance
+        self._last_quotes = defaultdict(lambda: None)
+        self._current_time = np.datetime64(0, "ns")
+        self._loader = {}
+
+        # - setup communication bus
+        self.set_communication_channel(bus := SimulatedCtrlChannel("databus", sentinel=(None, None, None)))
+        self.trading_service.set_communication_channel(bus)
+
+        # - simulated scheduler
+        self._scheduler = _SimulatedScheduler(bus, lambda: self.trading_service.time().item())
+
+        logger.info(f"SimulatedData.{exchange_id} initialized")
+
+    def subscribe(
+        self,
+        subscription_type: str,
+        instruments: List[Instrument],
+        timeframe: str | None = None,
+        nback: int = 0,
+        **kwargs,
+    ) -> bool:
+        units = kwargs.get("timestamp_units", "ns")
+
+        for instr in instruments:
+            _params: Dict[str, Any] = dict(
+                reader=self._reader,
+                instrument=instr,
+                preload_bars=nback,
+                timeframe=timeframe,
+            )
+
+            # - for ohlc data we need to restore ticks from OHLC bars
+            if "ohlc" in subscription_type:
+                _params["transformer"] = RestoreTicksFromOHLC(
+                    trades="trades" in subscription_type, spread=instr.min_tick, timestamp_units=units
+                )
+            elif "quote" in subscription_type:
+                _params["transformer"] = AsQuotes()
+
+            # - create loader for this instrument
+            self._loader[instr.symbol] = DataLoader(**_params)
+
+        return True
+
+    def run(self, start: str | pd.Timestamp, end: str | pd.Timestamp):
+        ds = []
+        for s, ld in self._loader.items():
+            data = ld.load(start, end)
+            ds.append(pd.Series({q.time: q for q in data}, name=s) if data else pd.Series(name=s))
```

Comment:
This will only work well for ohlc data, but if we want to simulate trades or orderbook updates on more than several weeks, we will get OOM. Maybe we should directly try to use the chunked version of data loading and process chunk by chunk? Or at least in the next PR.

Reply:
Yes, you are correct - I just made simplest version for testing.
Actually idea is to use same approach as in Qube - use chunked data where chunk size depends on type of data and timeframe. 

For example if we use OHLC 5Min bars we can load couple of years of data without OOM
for trades - iterate by 3-5 days etc